base/event.jl

```diff
@@ -343,3 +343,47 @@ function Timer(cb::Function, timeout::Real, repeat::Real=0.0)
     yieldto(waiter)
     return t
 end
+
+type TimeoutException <: Exception
+    msg::AbstractString
+
+    TimeoutException() = new("")
+    TimeoutException(msg) = new(msg)
+end
+
+"""
+    @set_timeout t ex [msg]
+
+Set a timeout of `t` seconds on the execution time of expression `ex`.
+Throws `TimeoutException` with optional error message `msg`
+if the timeout is triggered.
+"""
+macro set_timeout(timeout_s, ex, msgs...)
+    t_ = eval(timeout_s)
+
+    msg = isempty(msgs) ? ex : msgs[1]
+    if !isempty(msgs) && (isa(msg, Expr) || isa(msg, Symbol))
+        # message is an expression needing evaluating
+        msg = :(Main.Base.string($(esc(msg))))
+    elseif isdefined(Main, :Base) && isdefined(Main.Base, :string)
```

Comment:
Is this special case needed? Overhead of `string("string literal")` is extremely small.